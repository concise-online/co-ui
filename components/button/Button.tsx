import React, { ButtonHTMLAttributes } from 'react';
import classNames from 'classnames';
import { getPrefixCls } from '../utils/config';

const prefixCls = getPrefixCls('btn');

interface Props extends Omit<ButtonHTMLAttributes<HTMLButtonElement>, 'type'> {
  type?: 'primary' | 'success' | 'info' | 'warning' | 'danger';
  htmlType?: 'button' | 'submit' | 'reset';
  fill?: boolean;
}

function Button(props: Props) {
  const { children, className, htmlType, type, fill, ...btnProps } = props;
  const classes = classNames(prefixCls, className, {
    [`${prefixCls}-${type}`]: !!type,
    [`${prefixCls}-fill`]: fill,
  });

  return (
    <button
      type={htmlType}
      className={classes}
      {...btnProps}
    >
      <span>{children}</span>
    </button>
  );
}

export default Button;
