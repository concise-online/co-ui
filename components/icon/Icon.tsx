import React, { ComponentProps } from 'react';
import classNames from 'classnames';
import MdiIcon from '@mdi/react';
import icons, { register } from './icons';
import { getPrefixCls } from '../_utils/config';
import warning from '../_utils/warning';

const prefixCls = getPrefixCls('icon');

export interface Props extends Omit<ComponentProps<typeof MdiIcon>, 'path'> {
  name: string;
}

function Icon(props: Props) {
  const { name, className, ...iconProps } = props;

  if (!icons[name]) {
    warning(icons[name], 'icon', `icon ${name} is not exist`);
    return null;
  }

  return (
    <MdiIcon
      aria-label={`icon: ${name}`}
      path={icons[name]}
      size="1em"
      color="currentColor"
      className={classNames(prefixCls, className)}
      {...iconProps}
    />
  );
}

Icon.register = register;

export default Icon;
