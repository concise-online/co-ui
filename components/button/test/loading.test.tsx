import React from 'react';
import { shallow, render } from 'enzyme';
import Button from '../index';
import Icon from '../../icon';

describe('button loading feature', () => {
  it('render correctly', () => {
    const wrapper = render(<Button loading>Loading...</Button>);

    expect(wrapper).toMatchSnapshot();
  });

  it('load element is the first child', () => {
    const wrapper = shallow(<Button loading>Loading...</Button>);
    const loading = wrapper.childAt(0);

    expect(loading.type()).toBe(Icon);
    expect(loading.prop('name')).toBe('loading');
    expect(loading.prop('spin')).toBeDefined();
  });

  it('loading prop rewrite prefix icon prop', () => {
    const wrapper = shallow(<Button icon="face">Button</Button>);

    expect(wrapper.childAt(0).type()).toBe(Icon);
    expect(wrapper.childAt(0).prop('name')).toBe('face');
    wrapper.setProps({ loading: true });
    expect(wrapper.childAt(0).type()).toBe(Icon);
    expect(wrapper.childAt(0).prop('name')).toBe('loading');
  });
});
