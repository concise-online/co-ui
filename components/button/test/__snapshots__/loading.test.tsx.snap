// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`button loading feature render correctly 1`] = `
<button
  class="co-btn co-btn-loading"
>
  <svg
    aria-label="icon: loading"
    class="co-icon co-btn-icon"
    role="presentation"
    style="width:1em;height:1em"
    viewBox="0 0 24 24"
  >
    <g
      style="animation:spin linear 0.8s infinite;transform-origin:center"
    >
      <path
        d="M12,4V2C6.48,2 2,6.48 2,12H4C4,7.58 7.58,4 12,4Z"
        style="fill:currentColor"
      />
      <rect
        fill="transparent"
        height="24"
        width="24"
      />
    </g>
  </svg>
  <span>
    Loading...
  </span>
</button>
`;
